collections/collections.go

```diff
@@ -0,0 +1,104 @@
+	ErrNotFound = errors.New("collections: not found")
+	// ErrEncoding is returned when something fails during key or value encoding/decoding.
+	ErrEncoding = errors.New("collections: encoding error")
+)
+
+// StorageProvider is implemented by types
+// which provide a KVStore given a StoreKey.
+// It represents sdk.Context, it exists to
+// reduce dependencies.
+type StorageProvider interface {
+	// KVStore returns a KVStore given its StoreKey.
+	KVStore(key storetypes.StoreKey) storetypes.KVStore
+}
+
+// Prefix defines a segregation namespace
+// for specific collections objects.
+type Prefix struct {
+	raw []byte // TODO(testinginprod): maybe add a humanized prefix field?
+}
+
+// Bytes returns the raw Prefix bytes.
+func (n Prefix) Bytes() []byte { return n.raw }
+
+// NewPrefix returns a Prefix given the provided namespace identifier.
+// In the same module, no prefixes should share the same starting bytes
+// meaning that having two namespaces whose bytes representation is:
+// p1 := []byte("prefix")
+// p2 := []byte("prefix1")
+// yields to iterations of p1 overlapping over p2.
+// If a numeric prefix is provided, it must be between 0 and 255 (uint8).
+// If out of bounds this function will panic.
+// Reason for which this function is constrained to `int` instead of `uint8` is for
+// API ergonomics, golang's type inference will infer int properly but not uint8
+// meaning that developers would need to write NewPrefix(uint8(number)) for numeric
+// prefixes.
+func NewPrefix[T interface{ int | string | []byte }](identifier T) Prefix {
+	i := any(identifier)
+	var prefix []byte
+	switch c := i.(type) {
+	case int:
+		if c > math.MaxUint8 || c < 0 {
+			panic("invalid integer prefix value: must be between 0 and 255")
+		}
+		prefix = []byte{uint8(c)}
+	case string:
+		prefix = []byte(c)
+	case []byte:
+		identifierCopy := make([]byte, len(c))
+		copy(identifierCopy, c)
+		prefix = identifierCopy
+	}
+	return Prefix{raw: prefix}
+}
+
+// KeyCodec defines a generic interface which is implemented
+// by types that are capable of encoding and decoding collections keys.
+type KeyCodec[T any] interface {
+	// Encode writes the key bytes into the buffer. Returns the number of
+	// bytes written. The implementer must expect the buffer to be at least
+	// of length equal to Size(K) for all encodings.
+	// It must also return the number of written bytes which must be
+	// equal to Size(K) for all encodings not involving varints.
+	// In case of encodings involving varints then the returned
+	// number of written bytes is allowed to be smaller than Size(K).
+	Encode(buffer []byte, key T) (int, error)
+	// Decode reads from the provided bytes buffer to decode
+	// the key T. Returns the number of bytes read, the type T
+	// or an error in case of decoding failure.
+	Decode(buffer []byte) (int, T, error)
+	// Size returns the buffer size need to encode key T in binary format.
+	// The returned value must match what is computed by EncodeKey for all
```

Comment:
Where is EncodeKey defined?

Reply:
Whoops, doc issue... The API was renamed 10000times :)

Reply:
addressed, ty.